### ISSUE TITLE
oauth2-proxy issues with NLB + ALB setup in EKS cluster

### ISSUE THREAD

**Issue by kalinkostashki** [oauth2-proxy/oauth2-proxy#2353]
I have a monitoring stack of prometheus and grafana setup in a EKS cluster which I am trying to setup with AzureAD.
The message that I constantly receive is:
```
400 Bad Request
Request Header Or Cookie Too Large
```
How the flow should look like:

DNS look up -> Elastic IP attached to NLB -> NLB forwards to ALB -> ALB does TLS termination -> forwards to ingress-nginx controller -> looks up ingresses with configured auth-url -> get authenticated and redirected to the resource in question.

## Expected Behavior

1. Call Uri
2. Get redirected to Microsoft Login
3. After successful login get the resource being served by ingress
4. Be happy :)

## Current Behavior

1. Call Uri
2. Get redirected to Microsoft login
3. After sucessful auth get a 400 error
4. Look at ingress-nginx logs and cry:

```
10.18.1.245 - - [15/Dec/2023:20:40:56 +0000] "GET /oauth2/start?rd=%2Fgrafana%2F HTTP/1.1" 302 407 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" 827 0.001 [monitoring-oauth2-proxy-4180] [] 10.18.0.246:4180 407 0.001 302 f9ee6abb364ecee9b981ad37c6b4552a

10.18.1.245 - - [15/Dec/2023:20:41:14 +0000] "GET /oauth2/callback?code=0.ASEAGR7lCsgHS067bWSO5YQQ9BlIdFPuArBIlRG-OOh_sPEhAAA.AgABAAIAAAAmoFfGtYxvRrNriQdPKIZ-AgDs_wUA9P9crvbV-HnqEd5I6iW9rXLkLh3i1ZZj7-BcoW4OfOndCvfSwsrIaLLTpYP00dWAV2AwJSecx6PPs6tKyHi7pxS_MXGTbqGZJK279PDY7IdgxFAGNl8e5mMNZj6R2fCYrp9v-H8ifR7b7ugyKZUpWyi9UKAznKPFmMKK3LiB8FIAB_cdJqhtjfsxIELu3kSI1uQy6Zuicle_-fakBaWBhTRhuPpG0CHl7xC2HXMOdBBMFPhfZ4f3KMJ-nm-nRbUbU47E4TGUNtoQklZFuPPBZfxrMLJVwZzofQxzFDNgKb9WJHm7cjggR_mGyJFDbQW0oJtKdZ86-xPpKz8IDY23tW9hOkaFTD53wmS_PhV92DNFSCR1PbvhpoVnb4YJESyThZhgsVqAUSpO6xAzfvLpLVokDWWBkhyJxVAgSdEGuFZDlBWXxjdjXAp3NBawDO7OMwWc-JYPnxRslRVeT9k1gS8IpHbYiY89YDjp6BY8VZlss4AAKH35blLY-WL2pdVFOamszvcK3-ygDEE4c0G7hhhfX65zBXb4DSv6Gb91_HJJTRNVrxIKJJ4Km5IhfBkBZVjN7OAFVpK5FeLukabgD_2zTy8d7G8LJXuxuKbMBKGy4rM61tG3UT5ySLOvx8ofDzqqC6Lr0LX7O6Zs-f_U2yg4IFrR_E4fBD5X3lXDbMx9JWc75rpKfDU2dSV-olZn8grzNfsuVTe-zfbOVnV8b5QViOd3DwJL4gsNZlfMKFzCja3nBd3FHFlZxnUQs-FRytQSTJtT13geWmYPXRHy3k9hsc-sUBFItrDpqZSldsiOQ-w5x1pfCQ3Kd-UJXy2SbGv-eM-nG5DjBvwDYvt0prJfxRkWpdY5TGrO&state=bfNca6g1cxe7pOYIdzmUXwKKwMV7TqIRThlVVgPBIhI%3a%2fgrafana%2f&session_state=67368fde-b5af-4622-8b75-9f690a6dbf6e HTTP/1.1" 302 32 "https://login.microsoftonline.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" 2180 0.330 [monitoring-oauth2-proxy-4180] [] 10.18.0.246:4180 32 0.330 302 7e2d1528af83fe32c9ab4935867841d0

10.18.1.245 - - [15/Dec/2023:20:41:14 +0000] "GET /grafana/ HTTP/1.1" 400 628 "https://login.microsoftonline.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" 881 0.001 [] [] - - - - a9348548879102a3d013cac7182c2594

10.18.1.245 - - [15/Dec/2023:20:41:14 +0000] "GET /favicon.ico HTTP/1.1" 400 628 "https://my-domain.com/grafana/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" 760 0.001 [] [] - - - - b5e64f1db16f70fc07aab97e6b3629d4
```

As you can see auth is successful but upstream redirect isn't working.
Here are my ingress resource annotations:

```
    nginx.ingress.kubernetes.io/auth-signin: https://$host/oauth2/start?rd=$escaped_request_uri
    nginx.ingress.kubernetes.io/auth-url: http://oauth2-proxy.monitoring.svc.cluster.local:4180/oauth2/auth
    nginx.ingress.kubernetes.io/proxy-buffer-size: 16k
    nginx.ingress.kubernetes.io/rewrite-target: /$1
    nginx.ingress.kubernetes.io/service-upstream: "true"


```
## Possible Solution

What I did notice is that this works if there's no NLB... but due to my current setup this isn't something that I can work around.
So basically using a classical load balancer or just a plain ALB works.

## Steps to Reproduce (for bugs)

Install Grafana and Prometheus via [kube-prometheus](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) and bind it AzureAD App Registration(and I bet any other oauth provider)

## Context

I think I've summed it up for now. If more context is necessary I am happy to provide it.

## Your Environment

Oauth-Proxy version 7.4.0 (switching to 7.5.1 didn't help)
Kubernetes Ingress-Nginx: 1.9.4 (Helm Chart 4.8.3)

**Comment by kalinkostashki** [comment]
Ok, after some more trying around it appears that I found the fix.
You need to set large-client-header-buffers to a higher value. In my case:
`large-client-header-buffers: 4 32k`

This would be double the default set by nginx-ingress-controller.
Hope this helps someone :)
